Insert asset log (UI)

mark-prins opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #3273

# 👩🏻‍💻 What does this PR do? 
 <!-- Explain the changes you made, and why they're needed. Add a screenshot if you've made any UI changes!  -->
Creates a modal and wires up the insert api mutation.
Here's how the modal looks:

<img width="835" alt="image" src="https://github.com/msupply-foundation/open-msupply/assets/9192912/d27b3847-7d9c-49c0-8ca6-88ccfa132b2e">

With filtered reasons:

<img width="747" alt="image" src="https://github.com/msupply-foundation/open-msupply/assets/9192912/c33940a2-8ddc-4627-88f7-91e692129d33">


<img width="755" alt="image" src="https://github.com/msupply-foundation/open-msupply/assets/9192912/93057a9b-6346-4a6a-a137-f40fd2ac9ab7">

<img width="1512" alt="image" src="https://github.com/msupply-foundation/open-msupply/assets/9192912/583ade5e-af30-479a-afeb-90b47c6eea15">



# 🧪 How has/should this change been tested? 
<!-- Explain how to setup for testing here if it is not already obvious, and how you've tested this PR. -->
Try to create, check that cancel clears and closes the modal.
You'll need to check the database to see if the log is created successfully.

## 💌 Any notes for the reviewer?
<!-- eg. Do you have any specific questions for the reviewer? Is there a high risk/complicated change they should focus on? If there are any general areas of the codebase your changes might have have touched or could cause side effects to, mention them here.-->

## 📃 Documentation
<!-- Note down any areas which require documentation updates -->
Part of the CCEI feature

jmbrunskill commented:
Feels weird to have reason field with no options for me.
![image](https://github.com/msupply-foundation/open-msupply/assets/8287373/fe16cf2c-28e0-466d-a73e-9c0729e16a99)

@richardmoizeau what should we do here? Would it be better to have a reason like "No Reason Required?" in this case?

Also, if I'm doing a repair maintenance `Type` I feel like I should be entering a reason for the maintenance.
Like "Reason: Regular Scheduled Maintenance"

Not sure what the goals of collecting the reasons are?

jmbrunskill commented:
Had a bug where ok and next button is not enabled and not clear why...
![Kapture 2024-03-14 at 10 14 02](https://github.com/msupply-foundation/open-msupply/assets/8287373/4f6843cb-16a2-4a47-b2d1-9ce0f8013055)

jmbrunskill commented:
Functional Status doesn't show on the Summary Page, is that going to be fixed in a later PR?
![image](https://github.com/msupply-foundation/open-msupply/assets/8287373/ccf7551c-512a-483c-b70f-24d347e2f16a)

richardmoizeau commented:
> Feels weird to have reason field with no options for me. ![image](https://private-user-images.githubusercontent.com/8287373/312623278-fe16cf2c-28e0-466d-a73e-9c0729e16a99.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzNjQ3MTksIm5iZiI6MTcxMDM2NDQxOSwicGF0aCI6Ii84Mjg3MzczLzMxMjYyMzI3OC1mZTE2Y2YyYy0yOGUwLTQ2NmQtYTczZS05YzA3MjllMTZhOTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTNUMjExMzM5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjgyOTI4ODc2N2E2MWJkZGViN2M0OGRkZDE5YWUyN2FhMWM1MzgxYTVkZWM0ZmVhYzIxY2E3MTU5OWRiNjI0YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.MLW_z9-XCtCjU80YF0lNW61JN6HlpX2ZesWfbeGmpjw)
> 
> @richardmoizeau what should we do here? Would it be better to have a reason like "No Reason Required?" in this case?
> 
> Also, if I'm doing a repair maintenance `Type` I feel like I should be entering a reason for the maintenance. Like "Reason: Regular Scheduled Maintenance"
> 
> Not sure what the goals of collecting the reasons are?

reasons are used to complement the functional status (eg. functioning), not the type of record (eg. maintenance / repair). So should we rename "reason" to "status reason" for clarity ?

Reasons indicate why a status is what it is so when status is = "not functioning", you can specify a standard reason why. These reasons should be customisable (same as for the stocktake/inventory adjustment reasons): user can add/remove reasons for each status. "Status" on the other hand, should not be customisable. 

I agree with you, your wording is better when no reason is required: "No Reason Required?"

richardmoizeau commented:
> Had a bug where ok and next button is not enabled and not clear why... ![Kapture 2024-03-14 at 10 14 02](https://private-user-images.githubusercontent.com/8287373/312624597-4f6843cb-16a2-4a47-b2d1-9ce0f8013055.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzNjQ4MDAsIm5iZiI6MTcxMDM2NDUwMCwicGF0aCI6Ii84Mjg3MzczLzMxMjYyNDU5Ny00ZjY4NDNjYi0xNmEyLTRhNDctYjJkMS05Y2UwZjgwMTMwNTUuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTNUMjExNTAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDIxNTY3MDE5YzQxNGM4YTRhZDM0ZTA4MTJjNTI4Y2JiMDM5YjA5MzcwMWM2ZTljODhkZWNkMTdmNTEwNDFhYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Ki_M6XVDKuyx82ZYdqboVtabHmR9wuu7v_XcsUD_e0o) [ ![Kapture 2024-03-14 at 10 14 02](https://private-user-images.githubusercontent.com/8287373/312624597-4f6843cb-16a2-4a47-b2d1-9ce0f8013055.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzNjQ4MDAsIm5iZiI6MTcxMDM2NDUwMCwicGF0aCI6Ii84Mjg3MzczLzMxMjYyNDU5Ny00ZjY4NDNjYi0xNmEyLTRhNDctYjJkMS05Y2UwZjgwMTMwNTUuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTNUMjExNTAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDIxNTY3MDE5YzQxNGM4YTRhZDM0ZTA4MTJjNTI4Y2JiMDM5YjA5MzcwMWM2ZTljODhkZWNkMTdmNTEwNDFhYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Ki_M6XVDKuyx82ZYdqboVtabHmR9wuu7v_XcsUD_e0o) ](https://private-user-images.githubusercontent.com/8287373/312624597-4f6843cb-16a2-4a47-b2d1-9ce0f8013055.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzNjQ4MDAsIm5iZiI6MTcxMDM2NDUwMCwicGF0aCI6Ii84Mjg3MzczLzMxMjYyNDU5Ny00ZjY4NDNjYi0xNmEyLTRhNDctYjJkMS05Y2UwZjgwMTMwNTUuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTNUMjExNTAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDIxNTY3MDE5YzQxNGM4YTRhZDM0ZTA4MTJjNTI4Y2JiMDM5YjA5MzcwMWM2ZTljODhkZWNkMTdmNTEwNDFhYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Ki_M6XVDKuyx82ZYdqboVtabHmR9wuu7v_XcsUD_e0o) [ ](https://private-user-images.githubusercontent.com/8287373/312624597-4f6843cb-16a2-4a47-b2d1-9ce0f8013055.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTAzNjQ4MDAsIm5iZiI6MTcxMDM2NDUwMCwicGF0aCI6Ii84Mjg3MzczLzMxMjYyNDU5Ny00ZjY4NDNjYi0xNmEyLTRhNDctYjJkMS05Y2UwZjgwMTMwNTUuZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTNUMjExNTAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDIxNTY3MDE5YzQxNGM4YTRhZDM0ZTA4MTJjNTI4Y2JiMDM5YjA5MzcwMWM2ZTljODhkZWNkMTdmNTEwNDFhYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Ki_M6XVDKuyx82ZYdqboVtabHmR9wuu7v_XcsUD_e0o)

the only reason I see for this is that you have only one asset in your list OR that this asset is last record of your list.

adamdewey commented:
Yeah also agree with Richard that the reasons should be customisable

In that case could we just grey out the reason dropdown if there are none defined?

jmbrunskill commented:
> Yeah also agree with Richard that the reasons should be customisable
> 
> In that case could we just grey out the reason dropdown if there are none defined?

We can probably actually just hide the whole drop down in this case?

jmbrunskill commented:
> These reasons should be customisable (same as for the stocktake/inventory adjustment reasons): user can add/remove reasons for each status. "Status" on the other hand, should not be customisable.

Haha, we discussed this but didn't see any evidence that customisation was required, this will need to be a new ticket/issue as we'll need to create a process for managing these reasons, new configuration screens etc. Let's chat about this part separately.

mark-prins commented:
> Had a bug where ok and next button is not enabled and not clear why...

yeah, had that too. couldn't see why it happened though - so I gave up.

mark-prins commented:
> Functional Status doesn't show on the Summary Page, is that going to be fixed in a later PR?

Forgot about it, will pop it in this one 🙏

mark-prins commented:
> These reasons should be customisable

hold on!! this is a huge change! that was unclear from the reqs.
can we chat about this please?

jmbrunskill commented on client/packages/coldchain/src/Equipment/DetailView/UploadTab.tsx:
```diff
@@ -0,0 +1,10 @@
+export const UploadTab = ({ draft, value }: AssetLogPanel) => {
+  return (
+    <AssetLogPanel value={value} draft={draft}>
+      TODO: implement the upload panel
```
I wonder about things like this when commiting to develop? Ideally we should hide the tab maybe? I think it's ok at this stage, but as we get closer to a release...

mark-prins replied:
good point. have removed the wizard and next button etc, so the modal is now valid.

jmbrunskill commented on client/packages/coldchain/src/Equipment/api/hooks/log/useAssetLogInsert.ts:
```diff
@@ -0,0 +1,22 @@
+    async (log: Partial<InsertAssetLogInput>) => api.insertLog(log),
+    {
+      onSettled: assetId =>
+        queryClient.invalidateQueries(api.keys.logs(assetId ?? '')),
```
Do we need to invalidate the assets as well? As they have statuses from a loader which might be incorrect now?

mark-prins replied:
good point! have changed to use the base key and invalidated. now working

jmbrunskill reviewed (approved):
Looks good to me.

A couple of comments added but generally good to go I think.

jmbrunskill commented:
> yeah, had that too. couldn't see why it happened though - so I gave up.

I've recreated it several times, but also haven't figured out exactly why/when...

mark-prins commented:
Ok - have done the following:
* fixed the button enabled state
* added the log detail to the summary tab
* disabled the reason input when there are no options ( keeping elements visible but disabled is the preferred approach: 'we don't hide things' )
* cleared the reason when you change the status ( otherwise you can create invalid combinations )

bearing in mind this will have to change if we have customisable reason values, can we get this in?
